src/Authentication.Abstractions/AzureEnvironment.cs

```diff
@@ -131,18 +153,185 @@ public class AzureEnvironment : IAzureEnvironment
+        /// </summary>
+        /// <param name="armMetadata">ARM cloud metadata</param>
+        /// <returns></returns>
+        private static AzureEnvironment MapArmToAzureEnvironment(ArmMetadata armMetadata)
+        {
+            var azureEnvironment = new AzureEnvironment
+            {
+                Name = armMetadata.Name,
+                PublishSettingsFileUrl = GetPublishSettingsFileUrl(armMetadata.Name),
+                ServiceManagementUrl = armMetadata.Authentication.Audiences[0],
+                ResourceManagerUrl = armMetadata.ResourceManager,
+                ManagementPortalUrl = armMetadata.Portal,
+                ActiveDirectoryAuthority = armMetadata.Authentication.LoginEndpoint,
+                ActiveDirectoryServiceEndpointResourceId = armMetadata.Authentication.Audiences[0],
+                StorageEndpointSuffix = armMetadata.Suffixes.Storage,
+                GalleryUrl = armMetadata.Gallery,
+                SqlDatabaseDnsSuffix = armMetadata.Suffixes.SqlServerHostname,
+                GraphUrl = armMetadata.Graph,
+                TrafficManagerDnsSuffix = GetTrafficManagerDnsSuffix(armMetadata.Name),
+                AzureKeyVaultDnsSuffix = armMetadata.Suffixes.KeyVaultDns,
+                AzureKeyVaultServiceEndpointResourceId = GetKeyVaultServiceEndpointResourceId(armMetadata.Name),
+                AzureDataLakeAnalyticsCatalogAndJobEndpointSuffix = armMetadata.Suffixes.AzureDataLakeAnalyticsCatalogAndJob,
+                AzureDataLakeStoreFileSystemEndpointSuffix = armMetadata.Suffixes.AzureDataLakeStoreFileSystem,
+                DataLakeEndpointResourceId = armMetadata.ActiveDirectoryDataLake,
+                GraphEndpointResourceId = armMetadata.Graph,
+                BatchEndpointResourceId = armMetadata.Batch,
+                AdTenant = armMetadata.Authentication.Tenant
+            };
+
+            return azureEnvironment;
+        }
+
+        /// <summary>
+        /// Gets publish setting file URL for a cloud.
+        /// Note: Included for sake of parity, should remove if defunct.
+        /// </summary>
+        /// <param name="cloudName">Cloud name</param>
+        /// <returns></returns>
+        private static string GetPublishSettingsFileUrl(string cloudName)
+        {
+            switch (cloudName)
+            {
+                case EnvironmentName.AzureCloud:
+                    return AzureEnvironmentConstants.AzurePublishSettingsFileUrl;
+                case EnvironmentName.AzureChinaCloud:
+                    return AzureEnvironmentConstants.ChinaPublishSettingsFileUrl;
+                case EnvironmentName.AzureUSGovernment:
+                    return AzureEnvironmentConstants.USGovernmentPublishSettingsFileUrl;
+                case EnvironmentName.AzureGermanCloud:
+                    return AzureEnvironmentConstants.GermanPublishSettingsFileUrl;
+                default:
+                    return string.Empty;
+            }
+        }
+
+        // TODO: 
```

Comment:
empty todo

Reply:
Removed, thanks.

> @mdewan, could you please add some test cases to make sure there's breaking change introduced by arm metadata?

Even though ARM metadata has changed, or might change in the future, care has been taken to keep it backwards compatible - so a breaking change is highly unlikely. Secondly, unit tests will not guard against a breaking ARM metadata change, unless the test is updated when the breaking change occurs. Since the logic here simply deserializes ARM metadata, a unit test won't add value. If you think otherwise, please elaborate, and I am open to updating the PR, thanks.

Reply:
Refactored code, and added unit tests.